lib/Sema/CSDiagnostics.cpp

```diff
@@ -570,13 +570,26 @@ bool MissingConformanceFailure::diagnoseAsError() {
     }
   }
 
-  if (nonConformingType->isExistentialType()) {
-    auto diagnostic = diag::protocol_does_not_conform_objc;
-    if (nonConformingType->isObjCExistentialType())
-      diagnostic = diag::protocol_does_not_conform_static;
+  if (nonConformingType->isObjCExistentialType()) {
+    emitDiagnostic(anchor->getLoc(), diag::protocol_does_not_conform_static,
+                   nonConformingType, protocolType);
+    return true;
+  }
 
-    emitDiagnostic(anchor->getLoc(), diagnostic, nonConformingType,
-                   protocolType);
+  // Diagnose types that cannot conform to protocols.
+  Optional<StringRef> prohibitedTypeKind = None;
+  if (nonConformingType->is<AnyFunctionType>())
+    prohibitedTypeKind = "function";
+  else if (nonConformingType->is<TupleType>())
+    prohibitedTypeKind = "tuple";
+  else if (nonConformingType->isExistentialType())
+    prohibitedTypeKind = "protocol";
+  else if (nonConformingType->is<AnyMetatypeType>())
+    prohibitedTypeKind = "metatype";
```

Comment:
In my opinion we don't really need to spell this the kind of type here since they are all visually distinct e.g. metatype is going to have `.Type` at the end, protocol `.Protocol` etc.

Reply:
`.Type` and `.Protocol` are both metatypes, and the "protocol type" distinction here is for existential types. I agree that for functions, tuples, and metatypes, the type itself is enough to tell you what _kind_ of type it is, but I think it's useful for existential types because they look the same as struct/class/enum types. What do you think?

Reply:
How about “Non-nominal type %0 cannot conform to protocol %1”? Then you don’t need this extra logic.

Reply:
That was actually what I wrote originally, but "nominal" is a term only used in the compiler (it's not mentioned anywhere in TSPL) and many Swift programmers don't know what it means, especially beginners!

Reply:
I agree that it might make sense to have a special case for existentials something
`"{value of protocol |}%0type %1 cannot conform to %2; only struct/enum/class "
      "types can conform to protocols", (bool, Type, Type))`

Reply:
Yeah, that’s true. I remembered we have a similar diagnostic when extending a type like “Any” (non nominal type %0 cannot be extended) - perhaps it should be changed as well (separately, of course).

It would’ve been nice to have method that took a `TypeKind` and gave you back a string, similar to `DeclKind`/`DescriptiveDeclKind`, but I’m not sure if there’s a lot of existing code that has similar logic to the one you’ve got.

Reply:
I think it doesn't becomes when it comes to types most of the time (excluding existential types) it is self-evident what it is.